File: challenges/rate-limiter/typescript/starter/exercise.ts
import type { Clock } from "../../../../tooling/typescript/contracts.js";

// Implement the challenge contract here.
export class SlidingWindowLimiter {
  constructor(limit: number, windowMs: number, clock: Clock) {
    throw new Error("Not implemented: constructor");
  }

  allow(client: string): boolean {
    throw new Error("Not implemented: allow");
  }
}

File: challenges/rate-limiter/python/starter/exercise.py
"""Implement the contract in the challenge README. Reference code is separate."""
from tooling.python.contracts import Clock


class SlidingWindowLimiter:

    def __init__(self, limit: int, window_ms: int, clock: Clock):
        raise NotImplementedError('Not implemented: __init__')

    def allow(self, client: str) -> bool:
        raise NotImplementedError('Not implemented: allow')
